File: Models/level model/level_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:equatable/equatable.dart';

class LevelModel extends Equatable {
  final int levelNumber;
  final String name;
  final String pupilDescription;
  final String coachDescription;
  final String accessTier;
  final String? prerequisite;
  final bool isActive;
  final bool isPublished;
  final DateTime createdAt;
  final DateTime updatedAt;

  const LevelModel({
    required this.levelNumber,
    required this.name,
    required this.pupilDescription,
    required this.coachDescription,
    required this.accessTier,
    this.prerequisite,
    required this.isActive,
    required this.isPublished,
    required this.createdAt,
    required this.updatedAt,
  });

  factory LevelModel.create({
    required int levelNumber,
    required String name,
    required String pupilDescription,
    required String coachDescription,
    required String accessTier,
    String? prerequisite,
  }) {
    final now = DateTime.now();
    return LevelModel(
      levelNumber: levelNumber,
      name: name,
      pupilDescription: pupilDescription,
      coachDescription: coachDescription,
      accessTier: accessTier,
      prerequisite: prerequisite,
      isActive: true,
      isPublished: true,
      createdAt: now,
      updatedAt: now,
    );
  }

  // Factory constructor for Firestore DocumentSnapshot
  factory LevelModel.fromFirestoreDoc(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>?;
    if (data == null) {
      throw Exception('Document book bloc is null for level: ${doc.id}');
    }
    return LevelModel.fromJson(data);
  }

  // Factory constructor for Firestore book bloc (Map format)
  factory LevelModel.fromFirestore(Map<String, dynamic> data) {
    return LevelModel.fromJson(data);
  }

  // Factory constructor for Firestore QueryDocumentSnapshot
  factory LevelModel.fromFirestoreQuery(QueryDocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return LevelModel.fromJson(data);
  }

  // Firestore Serialization - Optimized for Firestore operations
  Map<String, dynamic> toFirestore() {
    final data = <String, dynamic>{
      'levelNumber': levelNumber,
      'name': name,
      'pupilDescription': pupilDescription,
      'coachDescription': coachDescription,
      'accessTier': accessTier,
      'isActive': isActive,
      'isPublished': isPublished,
      'createdAt': Timestamp.fromDate(createdAt),
      'updatedAt': Timestamp.fromDate(updatedAt),
    };

    // Add optional fields only if they have values
    if (prerequisite != null && prerequisite!.isNotEmpty) {
      data['prerequisite'] = prerequisite;
    }

    return data;
  }

  @override
  List<Object?> get props => [
    levelNumber,
    name,
    pupilDescription,
    coachDescription,
    accessTier,
    prerequisite,
    isActive,
    isPublished,
    createdAt,
    updatedAt,
  ];

  LevelModel copyWith({
    int? levelNumber,
    String? name,
    String? pupilDescription,
    String? coachDescription,
    String? accessTier,
    String? prerequisite,
    bool? isActive,
    bool? isPublished,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return LevelModel(
      levelNumber: levelNumber ?? this.levelNumber,
      name: name ?? this.name,
      pupilDescription: pupilDescription ?? this.pupilDescription,
      coachDescription: coachDescription ?? this.coachDescription,
      accessTier: accessTier ?? this.accessTier,
      prerequisite: prerequisite ?? this.prerequisite,
      isActive: isActive ?? this.isActive,
      isPublished: isPublished ?? this.isPublished,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // JSON Serialization (keeping original for backward compatibility)
  Map<String, dynamic> toJson() => {
    'levelNumber': levelNumber,
    'name': name,
    'pupilDescription': pupilDescription,
    'coachDescription': coachDescription,
    'accessTier': accessTier,
    'prerequisite': prerequisite,
    'isActive': isActive,
    'isPublished': isPublished,
    'createdAt': Timestamp.fromDate(createdAt),
    'updatedAt': Timestamp.fromDate(updatedAt),
  };

  // Helper method to safely parse DateTime from various formats
  static DateTime _parseDateTime(dynamic dateValue) {
    if (dateValue == null) {
      return DateTime.now();
    }

    if (dateValue is Timestamp) {
      return dateValue.toDate();
    }

    if (dateValue is String) {
      try {
        // Handle ISO 8601 strings (like from toIso8601String())
        return DateTime.parse(dateValue);
      } catch (e) {
        print('Error parsing date string: $dateValue, error: $e');
        return DateTime.now();
      }
    }

    if (dateValue is int) {
      // Assuming it's milliseconds since epoch
      return DateTime.fromMillisecondsSinceEpoch(dateValue);
    }

    print('Unexpected date type: ${dateValue.runtimeType}, value: $dateValue');
    return DateTime.now();
  }

  factory LevelModel.fromJson(Map<String, dynamic> json) {
    try {
      return LevelModel(
        levelNumber: (json['levelNumber'] is int)
            ? json['levelNumber'] as int
            : int.tryParse(json['levelNumber'].toString()) ?? 0,
        name: json['name']?.toString() ?? '',
        pupilDescription: json['pupilDescription']?.toString() ?? '',
        coachDescription: json['coachDescription']?.toString() ?? '',
        accessTier: json['accessTier']?.toString() ?? '',
        prerequisite: json['prerequisite']?.toString(),
        isActive: json['isActive'] is bool
            ? json['isActive'] as bool
            : (json['isActive'].toString().toLowerCase() == 'true'),
        isPublished: json['isPublished'] is bool
            ? json['isPublished'] as bool
            : (json['isPublished'].toString().toLowerCase() == 'true'),
        createdAt: _parseDateTime(json['createdAt']),
        updatedAt: _parseDateTime(json['updatedAt']),
      );
    } catch (e, stack) {
      print('❌ Error parsing LevelModel from JSON: $json');
      print('Error details: $e');
      print(stack);
      rethrow;
    }
  }

  @override
  String toString() =>
      'LevelModel(levelNumber: $levelNumber, name: $name, accessTier: $accessTier, isActive: $isActive, isPublished: $isPublished)';
}

File: src/model/LevelModel.ts
// src/models/LevelModel.ts
import { Timestamp, DocumentData, QueryDocumentSnapshot, DocumentSnapshot } from "firebase/firestore";

export interface ILevelModel {
  levelNumber: number;
  name: string;
  pupilDescription: string;
  coachDescription: string;
  accessTier: string;
  prerequisite?: string;
  isActive: boolean;
  isPublished: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export class LevelModel implements ILevelModel {
  levelNumber: number;
  name: string;
  pupilDescription: string;
  coachDescription: string;
  accessTier: string;
  prerequisite?: string;
  isActive: boolean;
  isPublished: boolean;
  createdAt: Date;
  updatedAt: Date;

  constructor(data: ILevelModel) {
    this.levelNumber = data.levelNumber;
    this.name = data.name;
    this.pupilDescription = data.pupilDescription;
    this.coachDescription = data.coachDescription;
    this.accessTier = data.accessTier;
    this.prerequisite = data.prerequisite;
    this.isActive = data.isActive;
    this.isPublished = data.isPublished;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }

  // Factory: create new LevelModel
  static create(params: {
    levelNumber: number;
    name: string;
    pupilDescription: string;
    coachDescription: string;
    accessTier: string;
    prerequisite?: string;
  }): LevelModel {
    const now = new Date();
    return new LevelModel({
      ...params,
      isActive: true,
      isPublished: true,
      createdAt: now,
      updatedAt: now,
    });
  }

  // Parse Firestore Date
  private static parseDate(value: any): Date {
    if (!value) return new Date();
    if (value instanceof Timestamp) return value.toDate();
    if (typeof value === "string") {
      try {
        return new Date(value);
      } catch {
        return new Date();
      }
    }
    if (typeof value === "number") return new Date(value);
    return new Date();
  }

  // From Firestore Document
  static fromFirestoreDoc(doc: DocumentSnapshot<DocumentData>): LevelModel {
    const data = doc.data();
    if (!data) {
      throw new Error(`Document is null for level: ${doc.id}`);
    }
    return LevelModel.fromJson(data);
  }

  static fromFirestoreQuery(doc: QueryDocumentSnapshot<DocumentData>): LevelModel {
    return LevelModel.fromJson(doc.data());
  }

  // From JSON/Plain object
  static fromJson(json: Record<string, any>): LevelModel {
    return new LevelModel({
      levelNumber:
        typeof json.levelNumber === "number"
          ? json.levelNumber
          : parseInt(json.levelNumber, 10) || 0,
      name: String(json.name ?? ""),
      pupilDescription: String(json.pupilDescription ?? ""),
      coachDescription: String(json.coachDescription ?? ""),
      accessTier: String(json.accessTier ?? ""),
      prerequisite: json.prerequisite ? String(json.prerequisite) : undefined,
      isActive:
        typeof json.isActive === "boolean"
          ? json.isActive
          : String(json.isActive).toLowerCase() === "true",
      isPublished:
        typeof json.isPublished === "boolean"
          ? json.isPublished
          : String(json.isPublished).toLowerCase() === "true",
      createdAt: LevelModel.parseDate(json.createdAt),
      updatedAt: LevelModel.parseDate(json.updatedAt),
    });
  }

  // To Firestore
  toFirestore(): Record<string, any> {
    const data: Record<string, any> = {
      levelNumber: this.levelNumber,
      name: this.name,
      pupilDescription: this.pupilDescription,
      coachDescription: this.coachDescription,
      accessTier: this.accessTier,
      isActive: this.isActive,
      isPublished: this.isPublished,
      createdAt: Timestamp.fromDate(this.createdAt),
      updatedAt: Timestamp.fromDate(this.updatedAt),
    };
    if (this.prerequisite && this.prerequisite.length > 0) {
      data.prerequisite = this.prerequisite;
    }
    return data;
  }

  // To JSON (for API/UI)
  toJson(): Record<string, any> {
    return {
      ...this.toFirestore(),
    };
  }

  copyWith(params: Partial<ILevelModel>): LevelModel {
    return new LevelModel({
      levelNumber: params.levelNumber ?? this.levelNumber,
      name: params.name ?? this.name,
      pupilDescription: params.pupilDescription ?? this.pupilDescription,
      coachDescription: params.coachDescription ?? this.coachDescription,
      accessTier: params.accessTier ?? this.accessTier,
      prerequisite: params.prerequisite ?? this.prerequisite,
      isActive: params.isActive ?? this.isActive,
      isPublished: params.isPublished ?? this.isPublished,
      createdAt: params.createdAt ?? this.createdAt,
      updatedAt: params.updatedAt ?? this.updatedAt,
    });
  }

  toString(): string {
    return `LevelModel(levelNumber: ${this.levelNumber}, name: ${this.name}, accessTier: ${this.accessTier}, isActive: ${this.isActive}, isPublished: ${this.isPublished})`;
  }
}
